core-primitives/stf-state-handler/src/state_handler.rs

```diff
@@ -25,144 +25,261 @@ use crate::{
+		&self,
+		shard: &ShardIdentifier,
+		state: &Repository::StateType,
+		state_hash: Repository::HashType,
+	) -> Result<()> {
+		let mut state_snapshots_lock =
+			self.state_snapshot_repository.write().map_err(|_| Error::LockPoisoning)?;
+
+		state_snapshots_lock.update(shard, state, state_hash)
+	}
 }
 
-impl<Repository, StateObserver> HandleState for StateHandler<Repository, StateObserver>
+impl<Repository, StateObserver, StateInitializer> HandleState
+	for StateHandler<Repository, StateObserver, StateInitializer>
 where
 	Repository: VersionedStateAccess,
+	Repository::StateType: SgxExternalitiesTrait + Hash<Repository::HashType>,
+	Repository::HashType: Copy,
 	StateObserver: UpdateState<Repository::StateType>,
+	StateInitializer: InitializeState<StateType = Repository::StateType>,
 {
-	type WriteLockPayload = Repository;
+	type WriteLockPayload = StatesMap<Repository::StateType, Repository::HashType>;
 	type StateT = Repository::StateType;
 	type HashType = Repository::HashType;
 
 	fn initialize_shard(&self, shard: ShardIdentifier) -> Result<Self::HashType> {
-		let mut state_write_lock =
-			self.state_snapshot_repository.write().map_err(|_| Error::LockPoisoning)?;
-		state_write_lock.initialize_new_shard(shard)
+		let initialized_state = self.state_initializer.initialize()?;
+		self.reset(initialized_state, &shard)
 	}
 
 	fn load(&self, shard: &ShardIdentifier) -> Result<Self::StateT> {
-		self.state_snapshot_repository
+		let state = self
+			.states_map_lock
 			.read()
 			.map_err(|_| Error::LockPoisoning)?
-			.load_latest(shard)
+			.get(shard)
+			.ok_or(Error::InvalidShard(*shard))?
+			.0
+			.clone();
+
+		Ok(state)
 	}
 
 	fn load_for_mutation(
 		&self,
 		shard: &ShardIdentifier,
 	) -> Result<(RwLockWriteGuard<'_, Self::WriteLockPayload>, Self::StateT)> {
-		let state_write_lock =
-			self.state_snapshot_repository.write().map_err(|_| Error::LockPoisoning)?;
-		let loaded_state = state_write_lock.load_latest(shard)?;
-		Ok((state_write_lock, loaded_state))
+		let state_write_lock = self.states_map_lock.write().map_err(|_| Error::LockPoisoning)?;
+		let state_clone = state_write_lock.get(shard).ok_or(Error::InvalidShard(*shard))?.0.clone();
+
+		Ok((state_write_lock, state_clone))
 	}
 
 	fn write_after_mutation(
 		&self,
-		state: Self::StateT,
+		mut state: Self::StateT,
 		mut state_lock: RwLockWriteGuard<'_, Self::WriteLockPayload>,
 		shard: &ShardIdentifier,
 	) -> Result<Self::HashType> {
+		state.prune_state_diff(); // Remove state diff before storing.
+		let state_hash = state.hash();
 		// We create a state copy here, in order to serve the state observer. This does not scale
 		// well and we will want a better solution in the future, maybe with #459.
-		let state_hash = state_lock.update(shard, state.clone())?;
+		state_lock.insert(*shard, (state.clone(), state_hash));
 		drop(state_lock); // Drop the write lock as early as possible.
 
+		self.update_state_snapshot(shard, &state, state_hash)?;
+
```

Comment:
When we update state, we write a snapshot to a file. We also update our observer (line below).